Task/DES-2001: Project links as observables (subscribe to project-user data).

duckonomy opened the pull request:
## Overview: ##

## PR Status: ##

* [X] Ready.
* [ ] Work in Progress.
* [ ] Hold.

## Related Jira tickets: ##

* [DES-2001](https://jira.tacc.utexas.edu/browse/DES-2001)

## Summary of Changes: ##
When saving/linking the map to a project, ensure that an observable project is also created.

## Testing Steps: ##
1. Save a map to a project and ensure that users are synced.(test with TACC-cloud/geoapi#65)


## UI Photos:

## Notes: ##

nathanfranklin reviewed (commented):
Looks good. 👍   I liked your ideas about changes to the modal that we discussed yesterday 💯

nathanfranklin reviewed (changes requested):
I can recheck when the backend PR is good to go but just wanted to leave a note of one things i saw 👍

duckonomy commented on src/app/components/modal-create-project/modal-create-project.component.ts:
```diff
@@ -68,12 +67,18 @@ export class ModalCreateProjectComponent implements OnInit, AfterContentChecked
+        this.projCreateForm.get('fileName').value
+      );
       this.close(project);
-    }, (err) => {
```
I've handled (1).

For (2), we have a separate error when `exportProject` (in `projectsService`) fails the request.

duckonomy commented:
### Changes
- [x] Refactored create project modal to clarify underlying functionality
- [x] Disallows modification of existing linked observable projects.
- [x] Removed user add in `users-panel` (Will further prune add-user related functionality in another PR. [Ticket](https://jira.tacc.utexas.edu/browse/DES-2083))
- [x] Removed Unused removed unused routes

duckonomy commented:
![Screenshot from 2021-12-22 14-26-03](https://user-images.githubusercontent.com/9425579/147151288-41bd3fd0-c810-4e13-bcbe-d634704ba7dd.png)

Newest changes address Tracy and Ellen's comments on usability.
1. By default, name is map name rather than map UUID (and automatically filled in when typing name).
2. All maps must be saved to a location for it to be created.
3. Clarify observable project option in the create project modal.
4. Indicate where we are saving to.
5. Remove "Save as" functionality for now
6. Prevent user for creating existing file
7. Assume that the user always syncs users when saving to projects.

nathanfranklin commented on src/app/components/modal-create-project/modal-create-project.component.ts:
```diff
@@ -60,43 +59,52 @@ export class ModalCreateProjectComponent implements OnInit, AfterContentChecked
       this.close(project);
     }, err => {
-      this.errorMessage = err.toString();
+      this.errorMessage = err.error.message;
```
We chatted about this section here 👍 .  As it doesn't involve staging/prod, we can ignore until a different PR or make more generic in this PR . whatever makes sense but just noting it as I wanted to pass along the info below 👇 

-----

For the geoapi-backend problem of not handling the backend exception and creating our expected response **when doing local dev**, it was unclear how to fix that. I thought i had previously but doesn't seem the case. its related to flask-restplus.  and instead of trying to fix that in the no longer supported `flask-restplus`, I think we need to just migrate to something that is supported. I created a ticket for that: https://jira.tacc.utexas.edu/browse/DES-2154.

nathanfranklin commented on src/app/components/modal-create-project/modal-create-project.component.ts:
```diff
@@ -60,43 +59,52 @@ export class ModalCreateProjectComponent implements OnInit, AfterContentChecked
     });
-    this.submitting = true;
+  }
+
+  fileAlreadyExists(fileName: string) {
+    return this.filesList.includes(`${fileName}.hazmapper`);
+  }
```
```suggestion
```

should we drop for now?

and when we do bring it back into the code base, we got to be careful as our file list isn't always complete for directories with large number of files  (e.g. as user scrolls, they get more of the file listing) .  i wonder if just doing a list for the single file would be a better approach

duckonomy replied:
I'll stash it locally and make a ticket 👍🏼

nathanfranklin commented on src/app/components/users-panel/users-panel.component.html:
```diff
@@ -103,12 +106,12 @@ <h6 id="property-title">Saved Location</h6>
-            <i class="fas fa-save"></i>  Save map link
-          </button>
-        </div>
+        <!-- <div *ngIf="!activeProject.system_id || (activeProject.system_id && !activeProject.system_id.includes('project'))" 
```
should we just drop?  or add a link to teh jira ticket here? then we could drop whenever we are certain that we aren't going to add this feature in.  (same applies fo `openExportProjectModal` method)

duckonomy replied:
I'll drop it for now.
Created https://jira.tacc.utexas.edu/browse/DES-2156.

nathanfranklin commented on src/app/components/users-panel/users-panel.component.ts:
```diff
@@ -136,18 +117,23 @@ export class UsersPanelComponent implements OnInit {
+      const pr = new ProjectRequest();
+      pr.project = this.activeProject;
+
+      this.projectsService.updateProject(pr);
```
check my comment on geoapi [here](https://github.com/TACC-Cloud/geoapi/pull/65/files#r706469782) as it kinda relates

as we are just changing the project name, i wonder if we need to delete and save the .hazmapper file in project service `updateProject` method

duckonomy replied:
I've mentioned this in the save-as ticket https://jira.tacc.utexas.edu/browse/DES-1929.
I think we could address it then.

nathanfranklin commented on src/app/components/users-panel/users-panel.component.ts:
```diff
@@ -136,18 +117,23 @@ export class UsersPanelComponent implements OnInit {
+      pr.project = this.activeProject;
+
+      this.projectsService.updateProject(pr);
+      // this.projectsService.updateProject(this.activeProject, pr);
```
drop

nathanfranklin commented on src/app/components/users-panel/users-panel.component.ts:
```diff
@@ -157,7 +143,12 @@ export class UsersPanelComponent implements OnInit {
+      const pr = new ProjectRequest();
+      pr.project = this.activeProject;
+
+      // this.projectsService.updateProject(this.activeProject, pr);
```
drop

nathanfranklin reviewed (commented):
Looks good to me.  👍  Just a few comments to consider.

I think the only thing that really needs to be considered is the updating of map metadata (like. map name) and what we should do when that happens.

duckonomy commented on src/app/components/modal-create-project/modal-create-project.component.html:
```diff
@@ -1,69 +1,64 @@
-                (systemSelection)="onSystemSelect($event)">
-              </app-file-browser>
+          </label>
+          <small><em>If a name is not given, file will be saved as the map name.</em></small>
```
I'll get rid of it 👍🏼

duckonomy commented on src/app/components/modal-create-project/modal-create-project.component.html:
```diff
@@ -1,69 +1,64 @@
+            <span style="padding-right: 10px">Sync Folder:</span>
+            <input formControlName="watchContent" type="checkbox">
+          </label>
+          <small><em>When enabled, files in this folder are automatically synced into the map periodically (1 hour)</em></small>
```
I think Troy from the RAPID team suggested that we have something to specify how long it takes. But I could omit it for now.

duckonomy commented on src/app/components/modal-create-project/modal-create-project.component.ts:
```diff
@@ -60,43 +59,52 @@ export class ModalCreateProjectComponent implements OnInit, AfterContentChecked
     const p = new Project();
+    const pr = new ProjectRequest();
+
     p.description = this.projCreateForm.get('description').value;
```
https://jira.tacc.utexas.edu/browse/DES-2157

nathanfranklin reviewed (approved):
LGTM 💯
